Update server validation logic.

kanishk98 opened the pull request:
<!--
Remove the fields that are not appropriate
Please include:
-->

**What's this PR do?**

1. Requests `zulip.ogg` as a fallback option in the case of HTTP 404 errors, preferring to use `getServerSettings` first. Fixes #573 and prevents `typo.zulipchat.com` from getting added. 

![Screenshot from 2019-05-25 03-12-23](https://user-images.githubusercontent.com/24617297/58361063-b3a4b700-7ea9-11e9-8198-6d68874cfc35.png)

2. Shows error message when trying to add a Zulip server with no organisations. Fixes #596.

![image](https://user-images.githubusercontent.com/24617297/58361132-2c0b7800-7eaa-11e9-890a-f9f18b64a4fe.png)


**You have tested this PR on:**
  - [ ] Windows
  - [x] Linux/Ubuntu
  - [ ] macOS

akashnimare commented:
@kanishk98 what happens in case of `abcd.zulip.org`?

priyank-p commented on app/renderer/js/utils/domain-util.js:
```diff
@@ -101,6 +101,48 @@ class DomainUtil {
+				const certErrorDetail = `The organization you're connecting to is either someone impersonating the Zulip server you entered, or the server you're trying to connect to is configured in an insecure way.
+				\nIf you have a valid certificate please add it from Settings>Organizations and try to add the organization again.
+				\nUnless you have a good reason to believe otherwise, you should not proceed.
+				\nYou can click here if you'd like to proceed with the connection.`;
```
Why is there a need to use template string literals and `\n`? This will mess up the formatting by adding double new lines.

priyank-p replied:
Also, I suggest you don't use template literal here since the indentation will show up. They are generally best for multiline string on top level.

kanishk98 replied:
> Why is there a need to use template string literals and \n?

I can't answer that - I just copied over the messages from the old method. 
I'll change this in a separate commit.

priyank-p replied:
Ah, can you try seeing if it affects the formatting of the dialog that shows up?

kanishk98 replied:
No, there's no change. 

Here's the dialog generated by the PR:

![mine](https://user-images.githubusercontent.com/24617297/58439700-28b6fd00-80f3-11e9-93df-c9c25816ed32.png)

Here's `master`:

![Screenshot from 2019-05-28 02-47-49](https://user-images.githubusercontent.com/24617297/58439709-2fde0b00-80f3-11e9-9877-7e0993b17a63.png)

The difference in the error message is because of the change in the first `request` made.

priyank-p replied:
Cool! The indentation will show up if you don't use `\n`.

priyank-p reviewed (commented):
Reading through it LGTM! Though we can however split the check for ogg into its own function like you did with for the cert error case since this whole logic seems to getting more complicated and hard to read through and follow.

priyank-p commented on app/renderer/js/utils/domain-util.js:
```diff
@@ -101,6 +101,48 @@ class DomainUtil {
 	}
 
+	checkCertError(domain, serverConf, error, silent) {
+		return new Promise((resolve, reject) => {
```
Make this a async function that way you don't need to return a promise.

priyank-p commented on app/renderer/js/utils/domain-util.js:
```diff
@@ -122,69 +160,59 @@ class DomainUtil {
 
-		return new Promise((resolve, reject) => {
-			request(checkDomain, (error, response) => {
+		return new Promise(async (resolve, reject) => {
```
Just make the `checkDomain` function async. You will to throw error where there is `Promise.reject()`.

priyank-p commented:
@kanishk98 if you interested in `async` conversion type of thing, we can use `request-promise` inside async functions and significantly reduced the callbacks we have and the code really readable and easy to follow, since we don't have to use lots of callback thus reduce the indentation and complex chaining we do.

kanishk98 commented:
> what happens in case of abcd.zulip.org?

Thanks for mentioning this! I missed out on a bug - I was checking for certificate errors even if the `error` returned from `getServerSettings()` was `undefined`, so `abcd.zulip.org` got stuck in that. 

I'll change the logic to only check for certificate errors when:
1. Either the `error` message includes `certificate`,
2. The domain is whitelisted.

kanishk98 commented:
> if you interested in async conversion type of thing, we can use request-promise inside async functions and significantly reduced the callbacks we have and the code really readable and easy to follow, since we don't have to use lots of callback thus reduce the indentation and complex chaining we do

@priyank-p that sounds like a good idea. The code's getting difficult to follow, and we should take this up along with our conversion to TypeScript if the others agree. 

Let's open a topic on czo to discuss this?

kanishk98 commented:
@akashnimare fixed the faulty test case. Here are the URLs I tried before updating this PR, all of which worked as expected:

1. `chat.zulip.org`
2. `abhigyank.zulipdev.org`
3. My own Zulip instance, with no orgs
4. `nucleus.zulipchat.com` (a real organisation)
5. `abcd.zulipchat.com`
6. `abcd.zulip.org`

@priyank-p I'm working on the `async-await` conversions right now, but I'm a little unsure about whether they should be part of this PR or become an independent change. 
I'll write them regardless, let me know what you think.

priyank-p commented:
You should do the async conversation here for the functions you changed in this PR in a seperate commit. Rest of them you should do it in a new PR.

priyank-p commented on app/renderer/js/utils/domain-util.js:
```diff
@@ -101,54 +102,80 @@ class DomainUtil {
+		try {
+			const response = await requestPromise(oggCheck);
+			if (response.statusCode === 200) {
+				// getting serverSettings has already failed
```
Why does this comment states something failed inside a 200 request condition?
Also I recommend checking range of good HTTP status i.e. 200 to but not including 400 since cached responses return 302 Not Modified.

kanishk98 replied:
> Why does this comment states something failed inside a 200 request condition?

I wanted to clarify why we're returning `serverConf` instead of `serverSettings`. Since we already tried retrieving the latter and got an error, we need to return `serverConf`. 

> I recommend checking range of good HTTP status i.e. 200 to but not including 400 since cached responses return 302 Not Modified.

That's a good catch, thanks. I'll update the PR.

abhigyank replied:
Can change the comment to explain this better and include the statement you mentioned.

kanishk98 replied:
@abhigyank marking this as resolved because we've decided to do away with the .ogg file check entirely, so there's no need to change the comments.

priyank-p commented on app/renderer/js/utils/domain-util.js:
```diff
@@ -101,54 +102,80 @@ class DomainUtil {
 	// ignoreCerts parameter helps in fetching server icon and
 	// other server details when user chooses to ignore certificate warnings
-	checkDomain(domain, ignoreCerts = false, silent = false) {
+	async checkDomain(domain, ignoreCerts = false, silent = false) {
```
I really liked reading the change for this function, this and other changes in this commit looks great. This version if far more readable.

kanishk98 replied:
Awesome, thanks! :)
Let me get a few more pressing issues out of the way and I'll begin work on converting all the code in the app to `async await`. Will let you know when I start.

priyank-p reviewed (commented):
@kanishk98 Thanks for doing the async migration. It looks great! Left one comment which might not be the change you made directly.

kanishk98 commented:
@akashnimare I've made the required changes. Please have a look whenever you find time.

kanishk98 commented:
As suggested by Tim in https://github.com/zulip/zulip-desktop/issues/573#issuecomment-498536198, I've removed the `.ogg` file check entirely. 

I also noticed that we were using a simple `alert()` for our error message in case the server the user connects with is invalid, so I changed that to a proper error box and changed the grammar a little. 
I also mentioned that we no longer Zulip server versions older than 1.6.

Here's what it looks like:

![image](https://user-images.githubusercontent.com/24617297/58935287-b4cbc300-878a-11e9-8c3f-d7b36d438b15.png)

kanishk98 commented:
@akashnimare whenever you have time, please take a look at this again. I think this is ready to merge.

akashnimare commented on app/renderer/js/utils/domain-util.js:
```diff
@@ -122,70 +157,37 @@ class DomainUtil {
-			});
-		});
+			} else {
+				const invalidZulipServerError = {
```
@priyank-p @abhigyank does it make sense to define this in a separate file and import it from there to make the code cleaner? Same with `certErrorMessage`, `certErrorDetail`. Code is getting cluttered because of these strings etc.

priyank-p replied:
Yeah, sounds good to me.

kanishk98 replied:
Cool. I'll add this file to the `resources` folder.

kanishk98 commented:
@akashnimare made a new file - `resources/messages.js`. We can use this to clean up our code as we go along. Currently it only has the error messages that this PR is concerned with.

akashnimare commented:
Maybe this could go to `\utils\`? cc - @abhigyank.

kanishk98 commented:
Hmm, I didn't think it was a good idea because the `utils` folder is inside `renderer` and I figured we could also use the same file for the main process. 
Regardless, that's a rather minor change at this stage.

abhigyank commented:
`resources` seem fine to me. Its not much of a `util` and regardless `util` is getting to many files too! :stuck_out_tongue:

abhigyank commented on app/renderer/js/pages/preference/new-server-form.js:
```diff
@@ -58,7 +59,7 @@ class NewServerForm extends BaseComponent {
 		}, errorMessage => {
 			this.$saveServerButton.children[0].innerHTML = 'Connect';
-			alert(errorMessage);
+			dialog.showErrorBox(errorMessage.title, errorMessage.content);
```
Have you tested this? Error returned from Line 145 doesn't have title or content keys in it.

kanishk98 replied:
Ah okay, I must've missed this while testing. 
Thanks, I'll check all error messages and make sure the title and content keys are present in each error message. 

Will edit this commit directly.

kanishk98 replied:
Okay, so pretty much the only clean way I know of to do this is to create new `Error` objects and then assign the properties `title` and `content` to them, which I feel is overkill for a feature that wasn't required in the first place (I just didn't like using `alert`). 

If you agree, then I can just drop this commit and go back to using `alert` with simple error strings. 

cc @akashnimare

kanishk98 replied:
(Or we could create message objects with `title` and `content` props, `stringify` them when throwing the error, and then parse them on the receiving end, but I think this is really messy).

akashnimare replied:
Alert sounds okay to me.

abhigyank reviewed (changes requested):
Some issue with the error message - 
![Screenshot from 2019-06-11 11-40-20](https://user-images.githubusercontent.com/20434085/59247673-bf300600-8c3d-11e9-85d9-bc8dce895da7.png)

akashnimare commented:
@kanishk98 can you take care of the above issue?

kanishk98 commented:
@akashnimare dropped the change from `alert` to `dialog` and fixed the error messages in `messages.js`. 
Let me know if any other changes are required here.

abhigyank reviewed (approved):
Wasn't able to test the `noOrgsError` but everything else works and looks fine to me. GTM.

vsvipul reviewed (approved):
LGTM. Good thought on creating a separate ```messages``` util.

kanishk98 committed:
domains: Fix server validation logic

Remove .ogg file check (supported only by very old servers). 
Reject domains with no organizations. 
Convert validation methods to async await
Add messages.js for returning error message strings.

akashnimare commented:
Merged. Thanks, everyone 🚀
